### ISSUE TITLE
SVG marker size

### ISSUE THREAD

**Issue by geraldo** [qgis2web/qgis2web#732]
I have an issue with the size of QGIS SVG markers: 

Using the default size works out great and marker appears correctly in the legend. To make it bigger on the map I change it's size from 4 to 6 Millimeters. But then it gets cut off on top and bottom. The marker is square but has margins on the left and right side.

Same behavior when using a rectangle marker without margins, like [this one](https://openclipart.org/download/192592/map-marker-blue.svg).

Is there something wrong with the attributes in this particular SVG marker or is this an expected behavior of qgis2web?

**Comment by tomchadwin** [comment]
Sounds like a bug to me. Does this happen with both Leaflet and OpenLayers exports?

**Comment by geraldo** [comment]
Right now I only can confirm this for Leaflet. Openlayers export seems to be broken right now on my system. I'll open an separate issue for that.

**Comment by geraldo** [comment]
This also happens on OpenLayers export with qgis2web 3.6.0.

**Comment by tomchadwin** [comment]
Is it using the width as the height?

**Comment by tomchadwin** [comment]
Your linked SVG didn't show up at all for me in OpenLayers. I then edited the SVG to make it square, and it worked.

**Comment by geraldo** [comment]
It works for you using, for example, the double marker size? Could you attach the maker to test it with my map?

It doesn't work so for me with, not with the linked marker but also not after making the marker square. It still gets cut off on top and bottom with this marker:
[map-marker-blue-square.svg.zip](https://github.com/tomchadwin/qgis2web/files/3019013/map-marker-blue-square.svg.zip)

Could the OS play a role in legend icons conversion to PNGs, I'm on Linux here.

**Comment by andreaordonselli** [comment]
The svg icon attached at the beginning of this discussion has a problem that is independent of the rectangular or square size. Openlayers can't read it.
The svg icons that you posted later always work in export, both with openlayers and with leaflets, and are clearly visible on the map at any size.
When you talk about "cut icon" do you mean in the legend like in the following image?

![image](https://github.com/qgis2web/qgis2web/assets/89784373/e8683400-0991-4d8c-bcf3-fb2f175c42ed)

**Comment by geraldo** [comment]
> When you talk about "cut icon" do you mean in the legend like in the following image?

Yes, that's exactly what happens when SVG icons are bigger than 4 mm: They get cropped in the legend. So would be nice to scale down in order to show icon entirely in the legend...

@andreaordonselli thanks for your great work on qgis2web

**Comment by andreaordonselli** [comment]
If you look at the icon in the QGIS legend you will notice that it also gets cropped when it gets large and qgis2web takes that icon. I'm thinking that in case of svg instead of taking the icon from the qgis legend I could directly bring the svg icon into the openlayer or leaflet legend. I'll look at the code and see what I can do.

**Comment by geraldo** [comment]
Using the original SVG icon in case of _Symbol layer type = SVG Marker_ sounds like a good solution. I could try to make an PR modifying _olLayerScripts.py_ and _leafletStyleScripts.py_ if you want...

**Comment by andreaordonselli** [comment]
I'm happy if you edit the files but I would prefer you to post the changes here because in my test environment those 2 files are already ahead of github.  Then, if I use your code, I will cite you in the release.  Thank you.

**Comment by geraldo** [comment]
After some tests I think we should do something like:

1. Check if symbol is of type `QgsSvgMarkerSymbolLayer`
2. If true output the marker as SVG
3. else output the marker as PNG as actually done

The code for OpenLayers could look something like that and should be inserted after [line 403 at olLayerScripts.py](https://github.com/qgis2web/qgis2web/blob/master/olLayerScripts.py#L403):

```
    if isinstance(renderer.symbol().symbolLayers()[0], QgsSvgMarkerSymbolLayer):
            svgFile = os.path.basename(renderer.symbol().symbolLayers()[0].path())
            layerCode += '''
                title: '<img style="width:16px;" src="styles/''' + svgFile + '''" /> %(name)s'
            });''' % {"icon": layerName,
                      "name": layer.name().replace("'", "\\'")}
    else:
            ...
```
I'm not quite sure if we just should check for first _symbolLayer_ or iterate over all of them...

If you are fine with the direction I'm taking, I could sketch it out also for Leaflet.

**Comment by andreaordonselli** [comment]
Congratulations, the code is perfect, please define max-width and max-height to 16px.
The ideal would be to replicate the code also in the def getLegend (line 418) so that everything also works with "QgsCategorizedSymbolRenderer" and with "QgsGraduatedSymbolRenderer".
If you don't have time let me know and I'll do it.
Thanks for now.

**Comment by andreaordonselli** [comment]
Here is the code I mean, can you test it and tell me what you think?
```python
    if isinstance(renderer, QgsSingleSymbolRenderer):
        if isinstance(renderer.symbol().symbolLayers()[0], QgsSvgMarkerSymbolLayer):
            svgFile = os.path.basename(renderer.symbol().symbolLayers()[0].path())
            layerCode += '''
                title: '<img style="max-width:16px; max-height:16px;" src="styles/''' + svgFile + '''" /> %(name)s'
            });''' % {"icon": layerName,
                      "name": layer.name().replace("'", "\\'")}
        else:
            layerCode += '''
                    title: '<img src="styles/legend/%(icon)s.png" /> %(name)s'
                });''' % {"icon": layerName,
                          "name": layer.name().replace("'", "\\'")}
    elif isinstance(renderer, QgsCategorizedSymbolRenderer):
        layerCode += getLegend(renderer.categories(), layer, layerName)
    elif isinstance(renderer, QgsGraduatedSymbolRenderer):
        layerCode += getLegend(renderer.ranges(), layer, layerName)
    else:
        layerCode += '''
                title: '%(name)s'
            });''' % {"name": layer.name().replace("'", "\\'")}
    return layerCode


def getLegend(subitems, layer, layerName):
    icons = ""
    for count, subitem in enumerate(subitems):
        if isinstance(subitem.symbol().symbolLayers()[0], QgsSvgMarkerSymbolLayer):
            svgFile = os.path.basename(subitem.symbol().symbolLayers()[0].path())
            icons += ("""\\
    <img style="max-width:16px; max-height:16px;" src="styles/%(icon)s" /> %(text)s<br />""" %
                      {"icon": svgFile, "text": subitem.label().replace("'", "\\'")})
        else:
            icons += ("""\\
    <img src="styles/legend/%(icon)s_%(count)s.png" /> %(text)s<br />""" %
                      {"icon": layerName, "count": count, "text": subitem.label().replace("'", "\\'")})
    
    legend = '''
    title: '%(name)s<br />%(icons)s'
        });''' % {"icons": icons, "name": layer.name().replace("'", "\\'")}
    return legend
```

**Comment by andreaordonselli** [comment]
I have just released version [3.20.0](https://github.com/qgis2web/qgis2web/releases/tag/v3.20.0), which fixes the issues discussed in this thread. Shortly, the update will be available within the QGIS plugin repository.

Thank you for using the plugin.

I would greatly appreciate it if you would consider donating a virtual coffee through the button on the homepage to support the development of qgis2web.

Best regards.

**Comment by geraldo** [comment]
Thanks for the fix and sorry for the later response, I didn't have time till now to check your code.

Everything works fine for point layers and SVG markers now do show up in the legend.

But I'm afraid we have to open this issue again as it doesn't work for `QgsCategorizedSymbolRenderer` and  `QgsGraduatedSymbolRenderer`. They actually don't use class `QgsSvgMarkerSymbolLayer` but instead `QgsSVGFillSymbolLayer`, which isn't implemented in _olLayerScripts.py_.

I'll post a PR with the code changes. Feel free to take only the parts you need in order not to mess up your internal code...

**Comment by andreaordonselli** [comment]
Thank you @geraldo 
why do you write to me that it doesn't work for you with QgsCategorizedSymbolRenderer and with QgsGraduatedSymbolRenderer?

two `elif ` are these, and they use `def getLegend` which I modified, in fact it works well for me even in categories or graduated style:

```python
elif isinstance(renderer, QgsCategorizedSymbolRenderer):
        layerCode += getLegend(renderer.categories(), layer, layerName)
    elif isinstance(renderer, QgsGraduatedSymbolRenderer):
        layerCode += getLegend(renderer.ranges(), layer, layerName)
```

```python
def getLegend(subitems, layer, layerName):
    icons = ""
    for count, subitem in enumerate(subitems):
        if isinstance(subitem.symbol().symbolLayers()[0], QgsSvgMarkerSymbolLayer):
            svgFile = os.path.basename(subitem.symbol().symbolLayers()[0].path())
            icons += ("""\\
    <img style="max-width:16px; max-height:16px;" src="styles/%(icon)s" /> %(text)s<br />""" %
                      {"icon": svgFile, "text": subitem.label().replace("'", "\\'")})
```

it seems to me that the only limitation is dictated by the fact that we are going to read the symbolLayers()[0] which is the first svg icon and if more than one is set the second one is discarded.

**Comment by geraldo** [comment]
Hi @andreaordonselli, I get errors with your code from v3.20:

1. _olLayerScripts.py_ throws an error as `sl` is of type `QgsSVGFillSymbolLayer` (and not `QgsSvgMarkerSymbolLayer` as expected) and so it does go into else.

>  File "/home/gerald/.local/share/QGIS/QGIS3/profiles/default/python/plugins/qgis2web/olStyleScripts.py", line 582, in getSymbolAsStyle sizeUnits = props["size_unit"] KeyError: 'size_unit' 

2. Once fixed this error the legend still does use the png image instead of svg, as again, `subitem.symbol().symbolLayers()[0]` is of type `QgsSVGFillSymbolLayer` (and not `QgsSvgMarkerSymbolLayer` as expected).

You can see my changes in [this commit](https://github.com/qgis2web/qgis2web/pull/1066/commits/8355f21f3f59bc696379c305ef7c3f441f78b00d) submitted as PR.

**Comment by andreaordonselli** [comment]
Does the error occur because you used a particular svg icon that is only read with QgsSVGFillSymbolLayer and not with QgsSvgMarkerSymbolLayer?
In any case I could configure both cases so as not to exclude anything, what do you think?

**Comment by geraldo** [comment]
It seems that _Singe Symbols_ do use  `QgsSvgMarkerSymbolLayer` to save _SVG Markers_ where as _Graduated_ and _Categorized_ do use `QgsSVGFillSymbolLayer` to save `SVG Fills`. So for what I tested it's not necessary to configure both, but I could miss something. So I'll send you my QGIS test project by mail so you can have a look at.

**Comment by andreaordonselli** [comment]
If you want to try my test file I configured the svgs from the standard qgis path. I'm exporting with the latest version of qgis2web with only the `QgsSvgMarkerSymbolLayer `configured

[qgis2web_test_qgz.zip](https://github.com/qgis2web/qgis2web/files/15306918/qgis2web_test_qgz.zip)

![image](https://github.com/qgis2web/qgis2web/assets/89784373/5e3276b2-6e64-4c37-9d97-31918ce9cf6f)

**Comment by andreaordonselli** [comment]
I saw your project, now I understand: you are not stylizing a point layer but a polygon layer, so you cannot choose a marker svg icon but you will choose a fill svg icon, which are not currently supported

![image](https://github.com/qgis2web/qgis2web/assets/89784373/67d2630b-fcc1-4336-a58b-7757995bbe2a)

In your tests with your code, is the map result good?

**Comment by geraldo** [comment]
You are right, sorry for that. I only had a look at the legend so I even didn't notice that the marker fill doesn't work.

So the thing is your code throws the error mentioned [here](https://github.com/qgis2web/qgis2web/issues/732#issuecomment-2109533145), my code at least avoids that error. But the better thing to do actually would be to draw at least the stroke of the polygon, actually that's what Leaflet does right now when it finds a (also not supported) marker fill.

So forget about the previous PR, I just submitted a new one which only fixes the error and draws the polygon with a default stroke for `QgsSVGFillSymbolLayer` in _olLayerScripts.py_.

**Comment by andreaordonselli** [comment]
Good morning Gerald, I've thought about it and in my opinion it's not good to make these micro updates just to allow a single project not to receive an error when exporting. Instead we should work to support the features and describe well in the wiki what is supported and what is not. I would like to go step by step. First solve the export of svg for point layers categorized or graduated styles (openalyers and leaflet) then proceed with lines and polygons. This would be good for qgis2web. The first thing to understand is whether openlayers and leaflet support this functionality and then program it in qgis2web. The help you have already provided is great but, in my opinion, we have to choose a direction and follow it. What do you think? Best regards

**Comment by geraldo** [comment]
I totally understand your point to go step by step. Right now I'm to busy, but I'll try to do some PRs on styling related to svgs in the future.

**Comment by andreaordonselli** [comment]
I retraced my steps, I simply modified this def. 
Now 40x40 icons are exported, not 16x16, and large svgs fit inside. 
Then with css I provide a maximum height and width of 16px in legend. This is how progressive numbering works too.
I'm working on exporting svg icons with progressive numbers to insert into the style, this is because if you use the same svg icon twice it will be exported only once and overwritten.

```
def singleSymbol(renderer, stylesFolder, layer_alpha, sln, legendFolder,
                 layers, feedback):
    symbol = renderer.symbol()
    (style, pattern, setPattern,
     useMapUnits) = getSymbolAsStyle(symbol, stylesFolder,
                                     layer_alpha, renderer, sln, layer,
                                     feedback)
    style = "var style = " + style
    legendIcon = QgsSymbolLayerUtils.symbolPreviewPixmap(
        symbol, QSize(40, 40))
    legendIcon.save(os.path.join(legendFolder, sln + ".png"))
    value = 'var value = ""'
    return (style, pattern, setPattern, value, useMapUnits)
```

**Comment by andreaordonselli** [comment]
I have just released version [3.21.0](https://github.com/qgis2web/qgis2web/releases/tag/v3.21.0), which fixes the issues discussed in this thread. The update is just available within the QGIS plugin repository.

Thank you for using the plugin.

I would greatly appreciate it if you would consider donating a virtual coffee through the button on the homepage to support the development of qgis2web.

Best regards.
